### PR TITLE
Correctly manage updates to RNGState

### PR DESCRIPTION
**Pull request by TeachRaccooon**
Making sure that random states are properly updated and passes around in functions. The convention is that every time RandBLAS generates matrices, we need to manually update random states. This process revealed a RandBLAS issue described [here](https://github.com/BallisticLA/RandBLAS/issues/57).
The API is now set up in such way that if a function uses randomness, then an RNGState appears as its final required argument.
Random states are mutated within functions.

I have also templated algorithm types in some of the test files.

### CODE REVIEW

File: RandBLAS
Comment (rileyjmurray) [inline RandBLAS:1]: Any idea why this is showing updates to RandBLAS in this PR?

**Comment by rileyjmurray** [comment]
@TeachRaccooon let's pass RNGState objects by reference with a ``const`` qualifier. That will reduce the chance that the user thinks we update RNGState automatically.

**Comment by rileyjmurray** [comment]
@TeachRaccooon I think I found a better way to manage the stack issues.

I tried instantiating the relevant structs used in testing with the ``new`` keyword so that they go on the heap instead of the stack. Then I dereferenced them before passing to functions that needed them. For example, here's what I have for ``TEST_F(TestQB, Polynomial_Decay_general1)`` after defining the boolean-valued subroutine parameters.
```
    auto all_data = new QBTestData<double>(m, n, k);
    auto all_algs = new algorithm_objects<double, r123::Philox4x32>(verbosity, cond_check, orth_check, p, passes_per_iteration, state);
    
    RandLAPACK::gen::mat_gen_info<double> m_info(m, n, RandLAPACK::gen::exponential);
    m_info.cond_num = 2025;
    m_info.rank = k;
    m_info.exponent = 2.0;
    RandLAPACK::gen::mat_gen<double, r123::Philox4x32>(m_info, (*all_data).A, state);

    svd_and_copy_computational_helper<double>(*all_data);
    test_QB2_low_exact_rank<double, r123::Philox4x32, algorithm_objects<double, r123::Philox4x32>>(block_sz, tol, *all_data, *all_algs);
```
The tests were failing on my machine (macOS) before this change, and then the pass after this change.

File: test/comps/test_qb.cc [hunk 1]
```diff
@@ -272,17 +275,20 @@ TEST_F(TestQB, Polynomial_Decay_general2)
...
-    QBTestData<double> all_data(m, n, k);
-    algorithm_objects<double, r123::Philox4x32> all_algs(verbosity, cond_check, orth_check, p, passes_per_iteration, state);
-    
-    RandLAPACK::gen::mat_gen_info<double> m_info(m, n, RandLAPACK::gen::polynomial);
-    m_info.cond_num = 6.7;
+    auto all_data = new QBTestData<double>(m, n, k);
+    auto all_algs = new algorithm_objects<double, r123::Philox4x32>(verbosity, cond_check, orth_check, p, passes_per_iteration, state);
+
+    RandLAPACK::gen::mat_gen_info<double> m_info(m, n, RandLAPACK::gen::exponential);
+    m_info.cond_num = 2025;
```
Comment (rileyjmurray) [inline test/comps/test_qb.cc:282]: Revert changes to these lines

Comment (TeachRaccooon) [reply test/comps/test_qb.cc:282]: Fixed

File: RandBLAS
Comment (TeachRaccooon) [reply RandBLAS:1]: I think the only place where we pass state into a function is in REVD, ortherwise state is a variable that belongs to a class. Are you suggesting that we should pass const state everywhere and then all those functions would also return states?

**Review by rileyjmurray — APPROVED** [review]

**Comment by TeachRaccooon** [comment]
@rileyjmurray This is not ready yet, wait.

**Review by rileyjmurray — APPROVED** [review]

**Comment by rileyjmurray** [comment]
Go ahead and merge once tests pass

**Comment by TeachRaccooon** [comment]
The only cases when something unorthodox gets returned form a function are SYPS and SYRF. I can update SYRF and REVD2 so that SYRF does not return a random state. SYPS is a more complicated case. 
I request either a full creative freedom to homogenize those tow with the rest of the API logic, or ask @rileyjmurray to update those himself.

**Comment by TeachRaccooon** [comment]
> The only cases when something unorthodox gets returned form a function are SYPS and SYRF. I can update SYRF and REVD2 so that SYRF does not return a random state. SYPS is a more complicated case. I request either a full creative freedom to homogenize those tow with the rest of the API logic, or ask @rileyjmurray to update those himself.

We can also wait with this until i decide to go through the entire randlapack and fix return types/error messages/orther stuff of that sort.

**Comment by rileyjmurray** [comment]
> > The only cases when something unorthodox gets returned form a function are SYPS and SYRF. I can update SYRF and REVD2 so that SYRF does not return a random state. SYPS is a more complicated case. I request either a full creative freedom to homogenize those tow with the rest of the API logic, or ask @rileyjmurray to update those himself.
> 
> We can also wait with this until i decide to go through the entire randlapack and fix return types/error messages/orther stuff of that sort.

I'll confer full creative freedom to you. You can manage that now or in a subsequent PR; your call.

**Comment by rileyjmurray** [comment]
@TeachRaccooon I'll let you be the one to squash-and-merge.

**Pull request merged**